### PR TITLE
Added route and handler to allow spawning from redirector

### PR DESCRIPTION
**Pull request by captainsafia**
This is a WIP PR for #4. I'll ping you on here @rgbkrk if I need any help.

### CODE REVIEW

File: redirector.py [hunk 1]
```diff
@@ -204,7 +222,7 @@ def main():
 
     settings = dict(
         stats=stats,
-        xsrf_cookies=True,
+        xsrf_cookies=False,
```
Comment (captainsafia) [inline redirector.py]: @rgbkrk — Do we need this set to `True` if we aren't using cookies for authentication?

Comment (rgbkrk) [inline redirector.py]: We can set this to False, tmpnb does too.

Comment (captainsafia) [inline redirector.py]: Yeah, that's why I was like:

wut

File: redirector.py [hunk 2]
```diff
@@ -93,6 +93,11 @@ def update_stats(stats):
             if host in stats:
                 stats[host] = data
 
+def fetch_api_spawn(host):
+    http_client = HTTPClient()
```
Comment (rgbkrk) [inline redirector.py]: I would use an `AsyncHTTPClient()` here, so it doesn't block the main event loop. After that, you'll need...

File: redirector.py [hunk 3]
```diff
@@ -93,6 +93,11 @@ def update_stats(stats):
...
+    req = HTTPRequest(host + "/api/spawn/", method="POST", body="")
+    response = http_client.fetch(req)
```
Comment (rgbkrk) [inline redirector.py]: ...to make this line be

```python
response = yield http_client.fetch(req)
```

**Comment by captainsafia** [comment]
So, whenever I `curl` on the `/api/spawn` endpoint, I get a `HTTP 405: Method Not Allowed` error despite the fact that I am sending using `POST` (`curl -X POST http://127.0.0.1:9000/api/spawn`). Not sure, why this is happening, complete stack-trace below.

```
Traceback (most recent call last):
  File "/Users/safia/anaconda3/lib/python3.4/site-packages/tornado/web.py", line 1415, in _execute
    result = yield result
  File "/Users/safia/anaconda3/lib/python3.4/site-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/Users/safia/anaconda3/lib/python3.4/site-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/Users/safia/anaconda3/lib/python3.4/site-packages/tornado/gen.py", line 876, in run
    yielded = self.gen.throw(*exc_info)
  File "redirector.py", line 180, in post
    response = yield http_client.fetch(request)
  File "/Users/safia/anaconda3/lib/python3.4/site-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/Users/safia/anaconda3/lib/python3.4/site-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
tornado.httpclient.HTTPError: HTTP 405: Method Not Allowed
```
I'm a confused cat and a frustrated ferret at the moment. I've tried laying out the request sending code in different ways but I don't think that's the problem.

**Comment by minrk** [comment]
@captainsafia I think it might be the trailing slash. Does it work if you do `/api/spawn/` instead? I opened jupyter/tmpnb#173 to make tmpnb behave the same way whether the slash is there or not.

The fact that the error is in `tornado.httpclient.HTTPError` shows (not super clearly) that the error is not in the request between curl and redirector, but in the second request from redirector to tmpnb. And it's the upstream tmpnb that has [`/api/spawn/`](https://github.com/jupyter/tmpnb/blob/1625cb528d6c03f71973fc148cad003f61de799a/orchestrate.py#L244) as the URL (with trailing slash).

**Comment by captainsafia** [comment]
@minrk — You are absolutely right! That's one pesky bug,  but good to note in the future. Thanks!

I updated the code and it works now.

File: redirector.py [hunk 4]
```diff
@@ -36,7 +36,7 @@
 from tornado import gen, web
 from tornado import ioloop
 
-from tornado.httpclient import HTTPRequest, HTTPError, AsyncHTTPClient
```
Comment (rgbkrk) [inline redirector.py]: You can delete the `HTTPClient` from the import now that you're using the `AsyncHTTPClient`

File: redirector.py [hunk 5]
```diff
@@ -172,6 +171,22 @@ def get(self):
     def stats(self):
         return self.settings['stats']
 
+class APISpawnHandler(RequestHandler):
+    @gen.coroutine
+    def post(self):
+        random_host = select_host(self.stats)
+        http_client = AsyncHTTPClient()
+        request = HTTPRequest(random_host + "/api/spawn/", method="POST", body="")
+        response = yield http_client.fetch(request)
```
Comment (rgbkrk) [inline redirector.py]: We should check the response (status code) to make sure we really have a server to hand back to them.

File: redirector.py [hunk 6]
```diff
@@ -172,6 +171,22 @@ def get(self):
...
+        data = json.loads(response.body.decode("utf-8"))
```
Comment (rgbkrk) [inline redirector.py]: Make sure to wrap this in a `try` block so we can deal with failure. I'm not sure whether we pass the error response back on to the user or what.

File: redirector.py [hunk 7]
```diff
@@ -172,6 +171,26 @@ def get(self):
...
+        random_host = select_host(self.stats)
+        http_client = AsyncHTTPClient()
+        request = HTTPRequest(random_host + "/api/spawn/", method="POST", body="")
+        try:
+            response = yield http_client.fetch(request)
+            data = json.loads(response.body.decode("utf-8"))
+            data["url"] = urljoin(random_host, data["url"])
+            self.write(data)
+        except Exception as e:
+            app_log.error("Failed to reach /api/spawn endpoint on %s: %s",
```
Comment (captainsafia) [inline redirector.py]: Should it write something on error? Perhaps an error JSON object?

Comment (rgbkrk) [inline redirector.py]: That sounds good until we have a real API definition for this behavior.

**Comment by captainsafia** [comment]
@rgbkrk — peep it!

**Comment by rgbkrk** [comment]
Looking good!

**Comment by rgbkrk** [comment]
In the next couple days I'll deploy this and see how thebe fairs with it.

**Pull request merged**